crates/bevy_ecs/src/schedule_v3/executor/multi_threaded.rs

```diff
@@ -0,0 +1,529 @@
+                // systems with zero dependencies
+                for (index, system_meta) in self.system_task_metadata.iter_mut().enumerate() {
+                    if system_meta.dependencies_total == 0 {
+                        self.ready_systems.set(index, true);
+                    }
+                }
+
+                // main loop
+                let world = SyncUnsafeCell::from_mut(world);
+                while self.completed_systems.count_ones(..) != self.completed_systems.len() {
+                    if !self.exclusive_running {
+                        self.spawn_system_tasks(scope, schedule, world);
+                    }
+
+                    if self.running_systems.count_ones(..) != 0 {
+                        // wait for systems to complete
+                        let index = self
+                            .receiver
+                            .recv()
+                            .await
+                            .unwrap_or_else(|error| unreachable!("{}", error));
+
+                        self.finish_system_and_signal_dependents(index);
+
+                        while let Ok(index) = self.receiver.try_recv() {
+                            self.finish_system_and_signal_dependents(index);
+                        }
+
+                        self.rebuild_active_access()
+                    }
+                }
+
+                // SAFETY: all systems have completed
+                let world = unsafe { &mut *world.get() };
+                Self::apply_system_buffers(&mut self.unapplied_systems, schedule, world);
+
+                debug_assert_eq!(self.ready_systems.count_ones(..), 0);
+                debug_assert_eq!(self.running_systems.count_ones(..), 0);
+                debug_assert_eq!(self.unapplied_systems.count_ones(..), 0);
+                self.active_access.clear();
+                self.completed_sets.clear();
+                self.completed_systems.clear();
+            };
+
+            #[cfg(feature = "trace")]
+            let runner_span = info_span!("schedule_task");
+            #[cfg(feature = "trace")]
+            let runner = runner.instrument(runner_span);
+            scope.spawn(runner);
+        });
+    }
+}
+
+impl MultiThreadedExecutor {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    fn spawn_system_tasks<'scope>(
+        &mut self,
+        scope: &Scope<'_, 'scope, ()>,
+        schedule: &'scope SystemSchedule,
+        world: &'scope SyncUnsafeCell<World>,
+    ) {
+        while let Some(system_index) = self
+            .ready_systems
+            .difference(&self.seen_ready_systems)
+            .next()
+        {
+            // skip systems we've already seen during this call
+            self.seen_ready_systems.set(system_index, true);
+
+            if self.system_task_metadata[system_index].is_exclusive {
+                {
+                    // SAFETY: no exclusive system running
+                    let world = unsafe { &*world.get() };
```

Comment:
Why are these two blocks separated?

Reply:
Just so that the `world` variable name is only overwritten in that scope. I need to get the pointer from the cell twice.